File: Binary Tree Preorder Traversal/binary-tree-preorder-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function preorderTraversal(root: TreeNode | null): number[] {
    const result: number[] = [];

    const helper = (node: TreeNode | null) => {
        if (!node) return;

        result.push(node.val);
        helper(node.left);
        helper(node.right);

        console.log(node.val);
    }
    helper(root);

    return result;
};

File: leetcode/a-easy/Binary Tree Preorder Traversal/Binary Tree Preorder Traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
    return preorderTraversalWith(root, []int{})
}

func preorderTraversalWith(node *TreeNode, current []int) []int {
    if node != nil {
        current = append(current, node.Val)
        current = preorderTraversalWith(node.Left, current)
        current = preorderTraversalWith(node.Right, current)
    }
    return current
}
